torch/_overrides.py

```diff
@@ -0,0 +1,235 @@
+        checked.
+    relevant_args : iterable
+        Iterable of arguments to check for __torch_function__ methods.
+    args : tuple
+        Arbitrary positional arguments originally passed into ``public_api``.
+    kwargs : tuple
+        Arbitrary keyword arguments originally passed into ``public_api``.
+
+    Returns
+    -------
+    Result from calling `implementation()` or an `__torch_function__`
+    method, as appropriate.
+
+    Raises
+    ------
+    TypeError : if no implementation is found.
+
+    """
+    # Check for __torch_function__ methods.
+    types, overloaded_args = get_overloaded_types_and_args(relevant_args)
+    # Short-cut for common cases: no overload or only Tensor overload
+    # (directly or with subclasses that do not override __torch_function__).
+    if (not overloaded_args or types == _TENSOR_ONLY):
+        return implementation(*args, **kwargs)
+
+    # Call overrides
+    for overloaded_arg in overloaded_args:
+        # Use `public_api` instead of `implemenation` so __torch_function__
+        # implementations can do equality/identity comparisons.
+        result = overloaded_arg.__torch_function__(
+            public_api.__name__, args, kwargs)
+
+        if result is not NotImplemented:
+            return result
+
+    func_name = '{}.{}'.format(public_api.__module__, public_api.__name__)
+    raise TypeError("no implementation found for '{}' on types that implement "
+                    '__torch_function__: {}'
+                    .format(func_name, list(map(type, overloaded_args))))
+
+
+ArgSpec = collections.namedtuple('ArgSpec', 'args varargs keywords defaults')
+
+def verify_matching_signatures(implementation, dispatcher):
+    """Verify that a dispatcher function has the right signature."""
+    implementation_spec = ArgSpec(*getargspec(implementation))
+    dispatcher_spec = ArgSpec(*getargspec(dispatcher))
+
+    if (implementation_spec.args != dispatcher_spec.args or
+        implementation_spec.varargs != dispatcher_spec.varargs or
+        implementation_spec.keywords != dispatcher_spec.keywords or
+        (bool(implementation_spec.defaults) !=
+         bool(dispatcher_spec.defaults)) or
+        (implementation_spec.defaults is not None and
+         len(implementation_spec.defaults) !=
+         len(dispatcher_spec.defaults))):
+        raise RuntimeError('implementation and dispatcher for %s have '
+                           'different function signatures' % implementation)
+
+
+_wrapped_func_source = textwrap.dedent("""
+    @functools.wraps(implementation)
+    def {name}(*args, **kwargs):
+        relevant_args = dispatcher(*args, **kwargs)
+        return implement_torch_function(
+            implementation, {name}, relevant_args, args, kwargs)
+    """)
+
+def torch_function_dispatch(dispatcher, module=None, verify=True,
+                            docs_from_dispatcher=False):
+    """Decorator for adding dispatch with the __torch_function__ protocol.
```

Comment:
Oh come on, give a little extra color to the message here :) Something like "If you define a function in Python and would like to permit user-defined tensor types with `__torch_function__` to override it, please apply this decorator on the function."